GpsDataCapturer/sharedlibrary/src/main/java/com/google/sharedlibrary/GpxFile.java

```diff
@@ -33,54 +36,105 @@ public static File createGpsDataFolder(Context context) {
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
+                context.getResources().getConfiguration().locale);
         return sdf.format(System.currentTimeMillis());
     }
 
-    public static void writeFileFooter(File gpxFile, Context context){
+    /**
+     * Create the xml header with version, creator and metadata
+     *
+     * @param formattedTime time of on location changed in format
+     * @return A header string
+     */
+    public static String writeFileHeader(String formattedTime) {
+        StringBuilder header = new StringBuilder();
+        header.append("<?xml version='1.0' encoding='UTF-8' ?>");
+        header.append("<gpx version=\"1.1\" creator=\"GpsDataCapturer " + BuildConfig.VERSION_CODE
+                + "\" ");
+        header.append("xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" ");
+        header.append("xmlns=\"http://www.topografix.com/GPX/1/1\" ");
+        header.append("xsi:schemaLocation=\"http://www.topografix.com/GPX/1/1 ");
+        header.append("http://www.topografix.com/GPX/1/1/gpx.xsd\">");
+        header.append("<metadata><time>").append(formattedTime).append("</time>");
+        header.append("<device>").append(Build.DEVICE).append("</device>");
+        header.append("<id>").append(Build.ID).append("</id>");
+        header.append("<manufacturer>").append(Build.MANUFACTURER).append("</manufacturer>");
+        header.append("<model>").append(Build.MODEL).append("</model></metadata>");
+        header.append("<trk>");
+        header.append("<trkseg>");
+        return header.toString();
+    }
+
+    /**
+     * Write the gpx file footer in xml format
+     *
+     * @param gpxFile the gpxFile
+     * @param context the context
+     */
+    public static void writeFileFooter(File gpxFile, Context context) {
         try {
             FileWriter fileWriter = new FileWriter(gpxFile, true);
-            BufferedWriter bufferedWriter = new BufferedWriter(fileWriter);
+//            BufferedWriter bufferedWriter = new BufferedWriter(fileWriter);
             Log.d(TAG, "Writing the xml footer.");
-            bufferedWriter.write("</trkseg></trk><time>" + getFormattedCurrentTime(context) + "</time></gpx>");
-            bufferedWriter.flush();
-            bufferedWriter.close();
+            fileWriter.write(
+                    "</trkseg></trk><time>" + getFormattedCurrentTime(context) + "</time></gpx>");
+//            bufferedWriter.flush();
+//            bufferedWriter.close();
             fileWriter.close();
             Log.i(TAG, "Finished writing to GPX file");
         } catch (IOException e) {
             Log.e(TAG, "Could not write the xml footer.", e);
         }
     }
-}
+
+    /**
+     * Reset the gpxFile to null
+     * @param gpxFile the gpxFile
+     */
+    public static void resetGpxFile(File gpxFile) {
```

Comment:
Similar here if make it instance method and gpxFile is not a parameter but member field.